test/distributed/org/apache/cassandra/distributed/test/BootstrapTest.java

```diff
@@ -18,85 +18,244 @@
+import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.service.StorageService;
 
+import static java.util.Arrays.*;
+import static org.apache.cassandra.distributed.action.GossipHelper.bootstrap;
+import static org.apache.cassandra.distributed.action.GossipHelper.decomission;
+import static org.apache.cassandra.distributed.action.GossipHelper.disseminateGossipState;
+import static org.apache.cassandra.distributed.action.GossipHelper.pullSchemaFrom;
+import static org.apache.cassandra.distributed.action.GossipHelper.statusToBootstrap;
 import static org.apache.cassandra.distributed.api.Feature.GOSSIP;
+import static org.apache.cassandra.distributed.api.Feature.NATIVE_PROTOCOL;
 import static org.apache.cassandra.distributed.api.Feature.NETWORK;
+import static org.apache.cassandra.distributed.shared.AssertUtils.assertRows;
+import static org.apache.cassandra.distributed.shared.AssertUtils.row;
 
-// TODO: this test should be removed after running in-jvm dtests is set up via the shared API repository
 public class BootstrapTest extends TestBaseImpl
 {
-
     @Test
     public void bootstrapTest() throws Throwable
     {
         int originalNodeCount = 2;
         int expandedNodeCount = originalNodeCount + 1;
-        Cluster.Builder builder = builder().withNodes(originalNodeCount)
-                                           .withTokenSupplier(TokenSupplier.evenlyDistributedTokens(expandedNodeCount))
-                                           .withNodeIdTopology(NetworkTopology.singleDcNetworkTopology(expandedNodeCount, "dc0", "rack0"))
-                                           .withConfig(config -> config.with(NETWORK, GOSSIP));
-
-        Map<Integer, Long> withBootstrap = null;
-        Map<Integer, Long> naturally = null;
-        try (Cluster cluster = builder.withNodes(originalNodeCount).start())
+
+        try (Cluster cluster = builder().withNodes(originalNodeCount)
+                                        .withTokenSupplier(TokenSupplier.evenlyDistributedTokens(expandedNodeCount))
+                                        .withNodeIdTopology(NetworkTopology.singleDcNetworkTopology(expandedNodeCount, "dc0", "rack0"))
+                                        .withConfig(config -> config.with(NETWORK, GOSSIP))
+                                        .start())
         {
-            populate(cluster);
+            populate(cluster,0, 100);
 
             IInstanceConfig config = cluster.newInstanceConfig();
-            config.set("auto_bootstrap", true);
-
-            cluster.bootstrap(config).startup();
+            IInvokableInstance newInstance = cluster.initialize(config);
+            withJoinRing(false, () -> newInstance.startup(cluster));
 
-            cluster.stream().forEach(instance -> {
-                instance.nodetool("cleanup", KEYSPACE, "tbl");
-            });
+            cluster.forEach(statusToBootstrap(newInstance));
+            cluster.run(asList(pullSchemaFrom(cluster.get(1)),
```

Comment:
comment was lost in the refactor, reposting

If you turn off joining the ring and disable autoboostrap (dtest default) then when you call `nodetool join` it will join the ring without bootstrapping (since autoboostrap=false). This test runs in that state, but then tries to bootstrap (which I don't think is currently possible), so feels like its testing a state we don't allow an operator to put us in?

If we add back auto_bootstrap=true then this logic can be replaced by 

```
IInstanceConfig config = cluster.newInstanceConfig();
config.set("auto_bootstrap", true);
IInvokableInstance newInstance = cluster.initialize(config);
withJoinRing(false, () -> newInstance.startup(cluster));
newInstance.nodetoolResult("join").asserts().success();
```

I took a stab and changed the test to do join in https://github.com/dcapwell/cassandra/commit/935f78b101484f6dfee473fd5375a31761f02b39

Reply:
Since the purpose of this patch is to introduce machinery to move cluster through arbitrary series of steps, I believe the test serves its purpose, since we'll need reliable runnable `bootstrap` and `pullSchemaFrom` for Harry. I've added the test you've implemented as `autoBootstrapTest`.

Reply:
> I've added the test you've implemented as autoBootstrapTest.

Works for me, fine with both versions.